File: Week1/problems/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Get credit card number
    long input = get_long("Number: ");

    // Find last number of credit card
    int lastDigit = input % 10;
    int secondLast = (input % 100 - lastDigit) / 10;
    // long temp = ((input % factor) - (input % (factor / 10))) / 10;

    long factor = 100;
    long sum = 0;

    // Go through every other number
    for (int i = 0; i < 8; i++)
    {
        // Formula to go through every other number
        long temp = ((input % factor) - (input % (factor / 10))) / (factor / 10);

        // Multiply each digit by 2
        temp = temp * 2;

        // Split up the product's characters
        if (temp >= 10)
        {
            temp = (temp % 10) + (temp % 100) / 10;
        }

        // Add the digits to the sum
        sum += temp;

        // Skip every other number
        factor = factor * 100;
    }

    // Find the remaining numbers and add it to sum
    factor = 10;
    for (int i = 0; i < 8; i++)
    {
        long temp = ((input % factor) - (input % (factor / 10))) / (factor / 10);

        // Split up the product's characters
        if (temp >= 10)
        {
            temp = (temp % 10) + (temp % 100) / 10;
        }

        // Add the digits to the sum
        sum += temp;

        // Skip every other number
        factor = factor * 100;
    }

    // Number is invalid if the sum doesn't end with a 0
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Determine which credit card company based on card digit length
    // American Express: 15
    if (input / 100000000000000 < 10 && input / 100000000000000 > 0)
    {
        int remainder = input / 10000000000000;
        if (remainder == 34 || remainder == 37)
        {
            printf("AMEX\n");
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 0;
        }
    }

    // Visa and Mastercard: 16
    if (input / 1000000000000000 < 10 && input / 1000000000000000 > 0)
    {
        int remainder = input / 100000000000000;
        if (remainder / 10 == 4)
        {
            printf("VISA\n");
            return 0;
        }

        if (remainder == 51 || remainder == 52 || remainder == 53 || remainder == 54 ||
            remainder == 55)
        {
            printf("MASTERCARD\n");
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 0;
        }
    }

    // Visa: 13
    if (input / 1000000000000 < 10 && input / 1000000000000 > 0)
    {
        printf("VISA\n");
        return 0;
    }

    else
    {
        printf("INVALID\n");
        return 0;
    }
}

File: Week6/problems/sentimental-credit/credit.py
# TODO

from cs50 import get_int
import math


def main():
    # Get user input and find first and last digits
    input = get_int("Number: ")

    lastDigit = math.floor(input % 10)
    secondLast = math.floor(math.floor(input % 100 - lastDigit) / 10)

    factor = 100
    sum = 0

    # Find every other number from input
    for i in range(8):
        # Formula to go through every other number
        temp = (
            math.floor(input % factor) - math.floor(input % math.floor(factor / 10))
        ) / math.floor(factor / 10)

        # Multiply each digit by 2
        temp = temp * 2

        # Split up the product's characters
        if temp >= 10:
            temp = math.floor(temp % 10) + math.floor(math.floor(temp % 100) / 10)

        # Add the digits to the sum
        sum += temp

        # Skip every other number
        factor = factor * 100

    # Find the remaining numbers and add it to sum
    factor = 10
    for i in range(8):
        temp = (
            math.floor(input % factor) - math.floor(input % math.floor(factor / 10))
        ) / math.floor(factor / 10)

        # Split up the product's characters
        if temp >= 10:
            temp = math.floor(temp % 10) + math.floor(math.floor(temp % 100) / 10)

        # Add the digits to the sum
        sum += temp

        # Skip every other number
        factor = factor * 100

    # Number is invalid if the sum doesn't end with a 0
    if math.floor(sum % 10) != 0:
        print("INVALID\n")
        return 0

    # Determine which credit card company based on card digit length
    # American Express: 15
    if (math.floor(input / 100000000000000) < 10) & (
        math.floor(input / 100000000000000) > 0
    ):
        remainder = math.floor(input / 10000000000000)
        if remainder == 34 or remainder == 37:
            print("AMEX\n")
            return 0

        else:
            print("INVALID\n")
            return 0

    # Visa and Mastercard: 16
    if (math.floor(input / 1000000000000000) < 10) & (
        math.floor(input / 1000000000000000) > 0
    ):
        remainder = math.floor(input / 100000000000000)
        if math.floor(remainder / 10) == 4:
            print("VISA\n")
            return 0

        if (
            (remainder == 51)
            or (remainder == 52)
            or (remainder == 53)
            or (remainder == 54)
            or (remainder == 55)
        ):
            print("MASTERCARD\n")
            return 0

        else:
            print("INVALID\n")
            return 0

    # Visa: 13
    if (math.floor(input / 1000000000000) < 10) & (
        math.floor(input / 1000000000000) > 0
    ):
        print("VISA\n")
        return 0

    else:
        print("INVALID\n")
        return 0


main()